### PR TITLE
Credentials refactor

### PR DESCRIPTION
**Pull request by abist**
Initial PR for re-routing credentials to use the native credentials (intercepted keytar node module) to save credentials. 

- [x] TODO: Discuss scope of Native Credentials Setting
- [ ] Test credential migration story on Linux (old to new system)

### CODE REVIEW

File: extensions/mssql/src/credentialstore/keychain.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+		try {
+			return await this.keytar.deletePassword(SERVICE_ID, credentialId);
+		} catch (e) {
+			// Ignore
+			Logger.error(`Deleting token failed: ${e}`);
+			return Promise.resolve(undefined);
+		}
+	}
+
+	// async tryMigrate(): Promise<string | null> {
```
Comment (abist) [inline extensions/mssql/src/credentialstore/keychain.ts:89]: WIP: Need to test migration on a Linux machine

File: extensions/mssql/src/utils.ts [hunk 2]
```diff
@@ -31,6 +31,16 @@ export function getAppDataPath() {
 	}
 }
 
+/**
+ * This function returns whether the native credential management system
+ * should be used instead of tools service
+ */
+export function nativeCredentialsEnabled() {
+	const linux: boolean = os.platform() === 'linux';
+	const nativeCredentialsEnabled: boolean = true; // add setting for this after review
```
Comment (abist) [inline extensions/mssql/src/utils.ts:40]: Need to discuss the scope of this setting with PM

File: src/sql/platform/credentials/common/credentialsService.ts [hunk 3]
```diff
@@ -8,7 +8,7 @@ import { createDecorator } from 'vs/platform/instantiation/common/instantiation'
 import * as azdata from 'azdata';
 import { Deferred } from 'sql/base/common/promise';
 
-export const SERVICE_ID = 'credentialsService';
+export const SERVICE_ID = 'sqlCredentialsService';
```
Comment (abist) [inline src/sql/platform/credentials/common/credentialsService.ts:11]: Had to change this because it clashes with VSCode's singleton `credentialService` decorator which is required to use the native credential service

File: extensions/mssql/src/sqlToolsServer.ts [hunk 4]
```diff
@@ -143,30 +147,34 @@ function generateHandleServerProviderEvent() {
...
+		SchemaCompareService.asFeature(context),
+		LanguageExtensionService.asFeature(context),
+		DacFxService.asFeature(context),
+		CmsService.asFeature(context),
+		SqlAssessmentService.asFeature(context),
+		NotebookConvertService.asFeature(context),
+		ProfilerFeature,
+		SqlMigrationService.asFeature(context)
+	];
+	if (nativeCredentialsEnabled()) {
```
Comment (abist) [inline extensions/mssql/src/sqlToolsServer.ts:167]: Add native credential service as a feature if the setting is set, otherwise use the server

File: extensions/mssql/src/credentialstore/keychain.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+	return undefined;
+}
+
+export type Keytar = {
+	getPassword: typeof keytarType['getPassword'];
+	setPassword: typeof keytarType['setPassword'];
+	deletePassword: typeof keytarType['deletePassword'];
+};
+
+// const OLD_SERVICE_ID = `${vscode.env.uriScheme}-microsoft.login`;
```
Comment (kburtram) [inline extensions/mssql/src/credentialstore/keychain.ts:31]: nit: dead code

File: extensions/mssql/src/credentialstore/keychain.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+			// likely when trying to refresh the token. Ignore the error since additional
+			// writes after the first one do not matter. Should actually be fixed upstream.
+			if (e.message === 'The specified item already exists in the keychain.') {
+				return;
+			}
+
+			const troubleshooting = localize('troubleshooting', "Troubleshooting Guide");
+			const result = await vscode.window.showErrorMessage(localize('keychainWriteError', "Writing login information to the keychain failed with error '{0}'.", e.message), troubleshooting);
+			if (result === troubleshooting) {
+				vscode.env.openExternal(vscode.Uri.parse('https://code.visualstudio.com/docs/editor/settings-sync#_troubleshooting-keychain-issues'));
```
Comment (kburtram) [inline extensions/mssql/src/credentialstore/keychain.ts:64]: is this link relevant to the errors we'll see in ADS?

File: extensions/mssql/src/credentialstore/logger.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+import * as vscode from 'vscode';
+
+type LogLevel = 'Info' | 'Error';
+
+class Log {
+	private output: vscode.OutputChannel;
+
+	constructor() {
+		this.output = vscode.window.createOutputChannel('SQL Credential Service');
```
Comment (kburtram) [inline extensions/mssql/src/credentialstore/logger.ts:14]: is this service only for `SQL`?  Is `Credential Service` already taken?

File: extensions/mssql/src/features.ts [hunk 8]
```diff
@@ -30,6 +32,68 @@ export class TelemetryFeature implements StaticFeature {
...
+		let saveCredential = (credentialId: string, password: string): Thenable<boolean> => {
+			return this.keychain.setPassword(credentialId, password).then(() => {
+				return this.keychain.getPassword(credentialId).then((savedPassword) => {
+					return savedPassword === password;
+				});
+			});
+		};
+
+		let deleteCredential = (credentialId: string): Thenable<boolean> => {
+			return this.keychain.deletePassword(credentialId).then((result) => { return result; });
```
Comment (kburtram) [inline extensions/mssql/src/features.ts:85]: do we need the `.then` block here? or can we return the promise coming back from `deletePassword` directly?

File: extensions/mssql/src/features.ts [hunk 9]
```diff
@@ -30,6 +32,68 @@ export class TelemetryFeature implements StaticFeature {
 	}
 }
 
+export class NativeCredentialsFeature extends SqlOpsFeature<any>  {
```
Comment (kburtram) [inline extensions/mssql/src/features.ts:35]: not sure `Native` is best name here (there term is a bit overloaded).  Maybe something like `KeytarCredentialsFeature` or `InternalCredentialsFeature` or something like that.

File: extensions/mssql/src/sqlToolsServer.ts [hunk 10]
```diff
@@ -87,11 +87,15 @@ export class SqlToolsServer {
 	}
 
 	private activateFeatures(context: AppContext): Promise<void> {
-		const credsStore = new CredentialStore(context.extensionContext.logPath, this.config);
 		const resourceProvider = new AzureResourceProvider(context.extensionContext.logPath, this.config);
-		this.disposables.push(credsStore);
 		this.disposables.push(resourceProvider);
-		return Promise.all([credsStore.start(), resourceProvider.start()]).then();
+		if (nativeCredentialsEnabled()) {
+			return Promise.all([resourceProvider.start()]).then();
```
Comment (kburtram) [inline extensions/mssql/src/sqlToolsServer.ts:93]: do we need `Promise.all` here?  it looks like there's just 1 promise so we can wait on it directly.

File: extensions/mssql/src/utils.ts (hunk 2 above)
Comment (kburtram) [inline extensions/mssql/src/utils.ts:40]: we should make this a user setting.  also we should default to false until we have the ability to move deal with existing credentials that need to be migrated.  if we aren't going to support credential migration, we should prompt the user if they have existing passwords and ask if the want to switch to this new mechanism.

**Review by kburtram — APPROVED** [review]
Need to switch the enabled default to `false` prior to merging, and properly handle scenario where user has existing credentials (either prompt and ask to enable, or migrate them over).

File: extensions/mssql/src/credentialstore/keychain.ts (hunk 6 above)
Comment (abist) [reply extensions/mssql/src/credentialstore/keychain.ts:64]: Yes, this is the standard resolution for macOS keychain and linux credential errors from the OS

**Comment by abist** [comment]
Closing this because I opened a new one here - https://github.com/microsoft/azuredatastudio/pull/14984
